contract/src/utils.rs

```diff
@@ -390,3 +365,336 @@ pub(crate) fn is_token_burned(token_identifier: &TokenIdentifier) -> bool {
+                &token_owner_item_key,
+            )
+            .unwrap_or_revert();
+            for token_identifier in owned_tokens_list.into_iter() {
+                let token_id = token_identifier.get_index().unwrap_or_revert();
+                let page_number = token_id / PAGE_SIZE;
+                let page_index = token_id % PAGE_SIZE;
+                let mut page_record = match storage::dictionary_get::<Vec<bool>>(
+                    all_owners_seed_uref,
+                    &token_owner_item_key,
+                )
+                .unwrap_or_revert()
+                {
+                    Some(page_record) => page_record,
+                    None => vec![false; page_record_width as usize],
+                };
+                let single_page_uref = get_uref(
+                    &format!("{}-{}", PAGE_DICTIONARY_PREFIX, page_number),
+                    NFTCoreError::MissingStorageUref,
+                    NFTCoreError::InvalidStorageUref,
+                );
+                let single_page_item_key =
+                    make_page_dictionary_item_key(&token_owner_key, page_number);
+                let _ = core::mem::replace(&mut page_record[page_number as usize], true);
+                storage::dictionary_put(all_owners_seed_uref, &token_owner_item_key, page_record);
+                let mut single_page = match storage::dictionary_get::<Vec<bool>>(
+                    single_page_uref,
+                    &single_page_item_key,
+                )
+                .unwrap_or_revert()
+                {
+                    None => vec![false; PAGE_SIZE as usize],
+                    Some(single_page) => single_page,
+                };
+                let is_already_marked_as_owned =
+                    core::mem::replace(&mut single_page[page_index as usize], true);
+                if is_already_marked_as_owned {
+                    runtime::revert(NFTCoreError::InvalidPageIndex)
+                }
+                storage::dictionary_put(single_page_uref, &single_page_item_key, single_page);
+                update_forward_and_reverse_token_trackers(token_id, token_identifier);
+                searched_token_ids.push(token_id)
+            }
+        }
+    }
+}
+
+pub(crate) fn should_break_up_owned_token_hashes(token_owner: Key) -> bool {
+    let all_owners_seed_uref = get_uref(
+        ALL_TOKEN_OWNERS,
+        NFTCoreError::MissingAllOwnersSeedURef,
+        NFTCoreError::InvalidAllOwnersSeedURef,
+    );
+    let is_absent_in_page_model = storage::dictionary_get::<Vec<bool>>(
+        all_owners_seed_uref,
+        &get_owned_tokens_dictionary_item_key(token_owner),
+    )
+    .unwrap_or_revert()
+    .is_none();
+    let is_present_in_owned_tokens = get_token_identifiers_from_dictionary(
+        &NFTIdentifierMode::Hash,
+        &get_owned_tokens_dictionary_item_key(token_owner),
+    )
+    .is_some();
+    is_present_in_owned_tokens && is_absent_in_page_model
+}
```

Comment:
change this to exit early

Reply:
Thinking about this, I don't know if exiting early is correct, if a user has migrated they still would retain the historical data, and the early exit would re-trigger the migration again. We should only migrate if both conditions are held, imo

Reply:
if is_present_in_owned_tokens is false you can return false, or if is_absent_in_page_model is false, you can return false so you should be able to evaluate the first condition and only do the work to check the second condition if the first is true